scripts/asr_language_modeling/ngram_lm/train_kenlm.py

```diff
@@ -23,144 +23,165 @@
     else:
-        with open(encoded_train_file, 'w', encoding='utf-8') as f:
-            for line in dataset:
-                f.write(f"{line}\n")
-
         discount_arg = ""
 
-    del model
-
     arpa_file = f"{args.kenlm_model_file}.tmp.arpa"
     """ LMPLZ ARGUMENT SETUP """
     kenlm_args = [
         os.path.join(args.kenlm_bin_path, 'lmplz'),
         "-o",
-        f"{args.ngram_length}",
-        "--text",
-        encoded_train_file,
+        str(args.ngram_length),
         "--arpa",
         arpa_file,
         discount_arg,
-    ]
+        "--prune",
+    ] + [str(n) for n in args.ngram_prune]
+
+    if args.cache_path:
+        if not os.path.exists(args.cache_path):
+            os.makedirs(args.cache_path, exist_ok=True)
+
+        """ DATASET SETUP """
+        encoded_train_files = []
+        for file_num, train_file in enumerate(train_paths):
+            logging.info(f"Encoding the train file '{train_file}' number {file_num+1} out of {len(train_paths)} ...")
+
+            cached_files = glob(os.path.join(args.cache_path, os.path.split(train_file)[1]) + "*")
+            encoded_train_file = os.path.join(args.cache_path, os.path.split(train_file)[1] + f"_{file_num}.tmp.txt")
+            if (
+                cached_files and cached_files[0] != encoded_train_file
+            ):  # cached_files exists but has another file name: f"_{file_num}.tmp.txt"
+                os.rename(cached_files[0], encoded_train_file)
+                logging.info("Rename", cached_files[0], "to", encoded_train_file)
+
+            encoded_train_files.append(encoded_train_file)
+
+        kenlm_utils.iter_files(
+            source_path=train_paths,
+            dest_path=encoded_train_files,
+            tokenizer=tokenizer,
+            encoding_level=encoding_level,
+            is_aggregate_tokenizer=is_aggregate_tokenizer,
+            text_processing_args=args.text_processing,
+            verbose=args.verbose,
+        )
 
-    ret = subprocess.run(kenlm_args, capture_output=False, text=True, stdout=sys.stdout, stderr=sys.stderr)
-    if ret.returncode != 0:
+        first_process_args = ["cat"] + encoded_train_files
+        first_process = subprocess.Popen(first_process_args, stdout=subprocess.PIPE, stderr=sys.stderr)
+
+        logging.info(f"Running lmplz command \n\n{' '.join(kenlm_args)}\n\n")
+        kenlm_p = subprocess.run(
+            kenlm_args,
+            stdin=first_process.stdout,
+            capture_output=False,
+            text=True,
+            stdout=sys.stdout,
+            stderr=sys.stderr,
+        )
+        first_process.wait()
+
+    else:
+        logging.info(f"Running lmplz command \n\n{' '.join(kenlm_args)}\n\n")
+        kenlm_p = subprocess.Popen(kenlm_args, stdout=sys.stdout, stdin=subprocess.PIPE, stderr=sys.stderr)
+
+        kenlm_utils.iter_files(
+            source_path=train_paths,
```

Comment:
When training a kenlm, do not do any text processing. This is English specific and does not work for other languages, but kenlm script training should be for all languages.

Reply:
done